streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredKeyValueStore.java

```diff
@@ -46,43 +46,90 @@
+public class MeteredKeyValueStore<K, V, OwnType extends MeteredKeyValueStore<K, V, OwnType>>
+    extends MeteredStore<K, V, MeteredKeyValueStore<K, V, OwnType>, KeyValueStore<Bytes, byte[]>>
     implements KeyValueStore<K, V> {
 
     final Serde<K> keySerde;
     final Serde<V> valueSerde;
     StateSerdes<K, V> serdes;
 
-    private final String metricsScope;
+    protected final String metricsScope;
     protected final Time time;
     protected Sensor putSensor;
-    private Sensor putIfAbsentSensor;
+    protected Sensor putIfAbsentSensor;
     protected Sensor getSensor;
-    private Sensor deleteSensor;
-    private Sensor putAllSensor;
-    private Sensor allSensor;
-    private Sensor rangeSensor;
-    private Sensor flushSensor;
-    private Sensor e2eLatencySensor;
-    private ProcessorContext context;
-    private StreamsMetricsImpl streamsMetrics;
-    private final String threadId;
-    private String taskId;
+    protected Sensor deleteSensor;
+    protected Sensor putAllSensor;
+    protected Sensor allSensor;
+    protected Sensor rangeSensor;
+    protected Sensor flushSensor;
+    protected Sensor e2eLatencySensor;
+    protected ProcessorContext context;
+    protected StreamsMetricsImpl streamsMetrics;
+    protected final String threadId;
+    protected String taskId;
 
     MeteredKeyValueStore(final KeyValueStore<Bytes, byte[]> inner,
                          final String metricsScope,
                          final Time time,
                          final Serde<K> keySerde,
                          final Serde<V> valueSerde) {
+        this(
+            inner,
+            metricsScope,
+            Thread.currentThread().getName(),
+            time != null ? time : Time.SYSTEM,
+            keySerde,
+            valueSerde
+        );
+    }
+
+    protected MeteredKeyValueStore(final KeyValueStore<Bytes, byte[]> inner,
+                         final String metricsScope,
+                         final String threadId,
+                         final Time time,
+                         final Serde<K> keySerde,
+                         final Serde<V> valueSerde) {
         super(inner);
         this.metricsScope = metricsScope;
-        threadId = Thread.currentThread().getName();
-        this.time = time != null ? time : Time.SYSTEM;
+        this.threadId = threadId;
+        this.time = time;
         this.keySerde = keySerde;
         this.valueSerde = valueSerde;
     }
 
+    public MeteredKeyValueStore<K, V, OwnType> reWrap(final KeyValueStore<Bytes, byte[]> inner) {
+        final MeteredKeyValueStore<K, V, OwnType> reWrapped = new MeteredKeyValueStore<>(
+            inner,
+            metricsScope,
+            threadId,
+            time,
+            keySerde,
+            valueSerde
+        );
+
+        copyInit(reWrapped);
+
+        return reWrapped;
+    }
+
+    protected void copyInit(final MeteredKeyValueStore<K, V, OwnType> reWrapped) {
```

Comment:
We're basically cloning an already initialized Metered store, so we need to copy over everything that gets set in init.